src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -152,40 +153,116 @@ internal Expression CreatePropertyNameExpression(IEdmEntityType elementType, IEd
+            {
+                foreach (ODataPathSegment segment in expandItem.PathToNavigationProperty)
+                {
+                    PropertySegment propertyAccessPathSegment = segment as PropertySegment;
+                    if (propertyAccessPathSegment != null)
+                    {
+                        IEdmProperty newProperty = propertyAccessPathSegment.Property;
+                        declaringType = newProperty.DeclaringType;
+                        break;
+                    }
+                    else
+                    {
+                        TypeSegment castSegment = segment as TypeSegment;
+                        if (castSegment != null)
+                        {
+                            continue;
+                        }
+
+                        break;
+                    }
+                }
+            }
 
+            Contract.Assert(declaringType != null, "Unstructured types cannot be projected.");
+            
             // derived property using cast
             if (elementType != declaringType)
             {
                 Type castType = EdmLibHelpers.GetClrType(declaringType, _model);
                 if (castType == null)
                 {
                     throw new ODataException(Error.Format(SRResources.MappingDoesNotContainResourceType,
-                        declaringType.FullName()));
+                        declaringType.FullTypeName()));
                 }
 
                 source = Expression.TypeAs(source, castType);
             }
 
+            Expression left = source;
             string propertyName = EdmLibHelpers.GetClrPropertyName(property, _model);
-            Expression propertyValue = Expression.Property(source, propertyName);
+            if (expandItem != null)
+            {
+                foreach (ODataPathSegment segment in expandItem.PathToNavigationProperty)
+                {
+                    string currentProperty = String.Empty;
+                    PropertySegment propertyAccessPathSegment =
+                         segment as PropertySegment;
+                    if (propertyAccessPathSegment != null)
+                    {
+                        IEdmProperty propertyInPath = propertyAccessPathSegment.Property;
+                        currentProperty = EdmLibHelpers.GetClrPropertyName(propertyInPath, _model);
+                    }
+                    else
+                    {
+                        TypeSegment typeSegment = segment as TypeSegment;
+                        if (typeSegment != null)
+                        {
+                            Type castType = EdmLibHelpers.GetClrType(typeSegment.EdmType, _model);
+                            left = Expression.TypeAs(left, castType);
+                        }
+                    }
+
+                    if (String.Compare(currentProperty, propertyName, StringComparison.Ordinal) == 0)
+                    {
+                        break;
+                    }
```

Comment:
Really?  what if the same property name is defined in my parent and nested type?  Don't you always want to build the expression for the entire PathToNavigationProperty?  Seems like better logic would be to handle the expandItem=null case as an else, so you don't have to try and stop one before the final property.